### DISCUSSION TITLE
chapter_data_structure/summary/

### DISCUSSION THREAD

**Discussion by giscus** [krahets/hello-algo#62] [Announcements]
# chapter_data_structure/summary/

Your first book to learn Data Structure And Algorithm.

https://www.hello-algo.com/chapter_data_structure/summary/

**Comment by shanghai-Jerry** [comment]
“所有数据结构都是由数组、链表或两者的组合实现的。”     道出了数据结果的本质

**Reply by shanghai-Jerry** [reply]
数据结构

**Comment by limuzi9806** [comment]
大佬！提个笔误？ “浮点数的编码由 1 位符号位、8 位指数位和 23 位分数为构成” 23位分数位

**Reply by krahets** [reply]
谢谢！！

**Comment by BlackCat-12** [comment]
好看，爱看，加油！

**Reply by krahets** [reply]
冲冲！

**Comment by zhuasdfg** [comment]
数据结构都是由数组和链表实现，让我大为震撼

**Comment by zzx1997QAQ** [comment]
我要是早半年看到就好了，感谢大佬！

**Comment by ChainWizard** [comment]
感谢大佬！

**Comment by advwei** [comment]
很清晰好理解，对于入门的学习者来说很友好了

**Comment by mobiusy** [comment]
每个章节的总结都很好，可以复查一下自己哪些还没理解

**Comment by NingWest01** [comment]
感谢分享

**Comment by Zhang-feixiang** [comment]
对小白太友好了  如果刚开始能知道这本书就好了

**Comment by FHMinyi** [comment]
我们老师上课太枯燥了，不如看这个有意思

**Comment by Crucifixion-Fxl** [comment]
请问如果是链式栈，是否就不是静态的数据结构了，这个貌似是可以动态添加的

**Reply by Autumn00air** [reply]
是的，链表的存储不是连续的，所以是动态的，而数组的存储空间必须是连续的空间，也就是事先要划定好。而有些语言的数组可以动态增加，其实是复制原本数组的数据，然后找一块更大的内存来存储这些数据，而链表的增加数据其实是可以跳跃式的找空的内存，所以是动态的。一般来说，基于数组实现的就是静态，基于链表实现的就是动态。（个人理解，欢迎指正~）

**Reply by Crucifixion-Fxl** [reply]
很不错的解释，我也认同你的想法！

**Comment by NgxuAnGMH** [comment]
这个教程真的泰酷辣！

**Q**：为什么哈希表同时包含线性数据结构和非线性数据结构？

哈希表底层是数组，而为了解决哈希冲突，我们可能会使用“链式地址”（后续“哈希冲突”章节会讲）：数组中每个桶指向一个链表，当链表长度超过一定阈值时，又可能被转化为树（通常为红黑树）。

从存储的角度来看，哈希表的底层是数组，其中每一个桶槽位可能包含一个值，也可能包含一个链表或一棵树。因此，哈希表可能同时包含线性数据结构（数组、链表）和非线性数据结构（树）。

**Comment by imxiaolong0804** [comment]
反码补码那个受益匪浅！！！

**Comment by DayDayYES** [comment]
数字是以补码的形式存储在计算机中，在加减法计算时，将减法转化为加法，引入补码就可以完成这些运算，而且解决了+-0的问题，用-0的对应的原码表示-128，这就是为什么一字节最小为-128.然后float同样为4字节，但是取值范围远大于int，因为float的采用不同的标准方式，第一比特是符号位，后八比特是指数位，最后23比特是分数位，虽然float取值范围大，但是其精度低，最后字符编码，我的python是UTF-8编码

**Comment by fdrama** [comment]
太厉害了，我不是计算机科班出身，对计算机如何表示数字的底层很好奇，以前看资料和教材，看到原码，反码，补码的概念其实有些云里雾里的。但是感到你这里的补数讲解，如醍醐灌顶，非常感谢您。

**Comment by HP-Patience** [comment]
2024.5.27打卡😊
之前对于原码，反码，补码的概念看过几次，没能很好地理解，就跳过了，没想到今天解决了

**Comment by PiFrancis** [comment]
原文：ASCII 码是最早出现的英文字符集，长度为 1 字节，共收录 127 个字符。

**Reply by PiFrancis** [reply]
应该是128

**Comment by zxcpyyya** [comment]
原码转补码的方法是“先取反后加 1”，那么补码转原码应该是逆运算“先减 1 后取反”，而补码转原码也一样可以通过“先取反后加 1”得到，这是为什么呢？
为什么例子中的0010取反变成了1101，前面不是说正数的反码就等于正数本身吗？有没有大佬解释一下

**Reply by YHDAO** [reply]
取反就是将二进制数字按位0变1，1变0。而取反的结果不一定是反码，正数的反码就是其本身，不需要取反。取反只是一种操作，而反码是一个数字编码方式。

**Reply by penghangtao** [reply]
上面的例子应该是只考虑负数的情况

**Comment by XiaosongChenXS** [comment]
我觉得关于这个问题“原码转补码的方法是“先取反后加 1”，那么补码转原码应该是逆运算“先减 1 后取反”，而补码转原码也一样可以通过“先取反后加 1”得到，这是为什么呢” 还可以这么解释：
首先，由于只有0和1这两个数字，所以取反自然是可逆：0 取反到 1 取反到 0；
其次，关键是在于加1的操作；
继续分析，分为两种情况
1） 假设原码最后1位是1，我们用x表示其它位数，也就是 xxxxxx1；
       我们用y表示x的取反，那么反码是 yyyyyy0；
       补码就是  yyyyyy1；也就是此时加1的操作是不影响反码的其它位数的；
      结论： 那从补码逆回去，补码 先-1再取反，相当于 1-1+1； 补码 先取反再+1，相当于 1-1+1；他们是对称的
                 也就是从原码 1变成0再变成1（取反再加1这个过程是对称的）
                并且无论是加1还是减1都不影响 ，yyyyyy取反后是xxxxxx；
2） 假设假设原码最后1位不是1，我们就找到从最后1位起为0的数值，假设是 xxxxxx1000;
       反码就是 yyyyyy0001；
       补码就是 yyyyyy1000; 
       发现没， 1000 变成 0001 变成 1000 这个过程是对称的 原码的1000 通过取反+1变成了 反码1000
      自然反码的 1000 可以通过取反+1变成了 原码1000

**Reply by XiaosongChenXS** [reply]
第二点有笔误，校正为：
假设假设原码最后1位不是1，我们就找到从最后1位起为0的数值，假设是 xxxxxx1000;
反码就是 yyyyyy0111；
补码就是 yyyyyy1000;
发现没， 1000 变成 0111 变成 1000 这个过程是对称的 原码的1000 通过取反+1变成了 反码1000
自然反码的 1000 可以通过取反+1变成了 原码1000

**Reply by shuxing-Long** [reply]
谢谢，看到你这里的“对称”,就明白了，在二进制中 加一和取反是等价的。

**Comment by wzhuo2022** [comment]
有个问题，正数的反码，补码不是都不变嘛，所以这里的例子是否有问题？

**Reply by XiaosongChenXS** [reply]
你可以理解为这个例子就是为负数的情况做说明的，正数都不变了，所以不在讨论范围内

**Comment by lastVigo** [comment]
这章的每一节都能展开讲半天。

**Comment by rui-project** [comment]
从下一章开始就要进入多练模式啦！加油呀！

**Comment by tomorrowthief** [comment]
所以补码，为啥一开始是取反加1

**Reply by Lovi-Dzy** [reply]
取反其实可以理解为换了一个基准，因为7位二进制取反之后，在十进制里的操作相当于是拿2的八次方-1减去这个数的绝对值，也就是说本来的负数是相对于0小多少，考虑操作可行性的问题，然后它转化成了负x就是间接表示成比（2的8次方-1）少x的数，这样加上一个正数之后，如果没超过2的八次方，那还是负数，结果再次取反就可以知道这个结果比2的七次方少多少，然后就是负多少，如果结果超过2的八次方就表示结果在10进制下是正数，循环进位处理之后，就可以知道这个数比2的七次方大多少，也就是正多少，结果就是正数，正好2的八次方就相当于0。补位加一的话主要是解决0的矛盾。我是这么理解的，你看下有没有道理

**Comment by SuperZhangRui** [comment]
没看原码补码反码的内容， 以后再看 继续加油

**Comment by hanzhe-one** [comment]
后续看

**Comment by linhf123** [comment]
先减 1 取反. 和先取反在加 1 在数学上就是等价的？就是一个先加后乘，一个先乘后加
（100+1）* -1 = 100 * -1 + 1*-1 = -101

**Comment by xiaozhuge123** [comment]
day02

**Comment by solink7** [comment]
# 打卡 2025/6/7
wow

**Comment by Ivy129** [comment]
打卡 2025/6/10
这个地方对于补码、原码、反码有了更加深刻的认识和理解

**Comment by Lovi-Dzy** [comment]
#打卡2025/7/10
---
高三毕业牲的暑假！
加油！(ง•̀_•́)ง

**Reply by nth1ng** [reply]
太卷了吧

**Reply by nanmingjushi** [reply]
牛

**Reply by swiftzhao** [reply]
牲是自嘲吗？哈哈，第一次听说

**Comment by zhuchentong** [comment]
坚持学习，打卡 2025.08.04

**Comment by xucy-code** [comment]
最重要的一句话 
（因为恒有 原码 + 反码 = 1111）

**Reply by duolingo1299** [reply]
还真是，1111+1刚好是=10000（不考虑进位的话恰好是0000）

**Comment by wangwangyou** [comment]
补码可以看作一个工程上的实践，很好的解决了0有两种表示的问题
对于反码来说，0有两种表示方式
正数 0 0000
负数 1 0000
正数的反码 0 0000
负数的反码 1 1111
但是如果使用补码的话
正数0的补码是 0 0000
负数0的补码是 0 0000 (丢弃最高位)

**Comment by z12jw** [comment]
在十进制中0.1，是个精确小数，转换为二进制后是无限小数，会截断到23位，所以说浮点数的编码牺牲了精度
